File: backend/app/schemas/release.py
from datetime import datetime
from enum import StrEnum
from uuid import UUID

from pydantic import (
    AnyHttpUrl,
    BaseModel,
    ConfigDict,
    Field,
    JsonValue,
    field_validator,
    model_validator,
)


class DashboardCardSetting(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1, max_length=50)
    visible: bool = True


class DashboardSettingWrite(BaseModel):
    model_config = ConfigDict(extra="forbid")

    cards: list[DashboardCardSetting] = Field(min_length=1, max_length=20)

    @model_validator(mode="after")
    def unique_cards(self) -> "DashboardSettingWrite":
        identifiers = [item.id for item in self.cards]
        if len(identifiers) != len(set(identifiers)):
            raise ValueError("Dashboard-Kacheln dürfen nicht doppelt vorkommen")
        return self


class DashboardSettingRead(DashboardSettingWrite):
    updated_at: datetime


class PortNameScheme(StrEnum):
    NUMERIC = "numeric"
    GIGABIT = "gigabit"
    ETHERNET = "ethernet"


class PortGroupWrite(BaseModel):
    model_config = ConfigDict(extra="forbid")

    group: str = Field(pattern="^(copper|sfp|sfp_plus|uplink)$")
    count: int = Field(ge=0, le=256)
    scheme: PortNameScheme = PortNameScheme.NUMERIC
    start: int = Field(default=1, ge=1, le=4096)
    speed_mbps: int | None = Field(default=None, ge=1, le=1_000_000)
    poe_capable: bool = False


class PortGenerationWrite(BaseModel):
    model_config = ConfigDict(extra="forbid")

    groups: list[PortGroupWrite] = Field(min_length=1, max_length=4)


class PortGenerationPreview(BaseModel):
    model_config = ConfigDict(extra="forbid")

    device_id: UUID
    existing_names: list[str]
    create_names: list[str]
    unchanged_names: list[str]
    requested_total: int


class PortGenerationResult(PortGenerationPreview):
    created: int


class NetworkPathNode(BaseModel):
    model_config = ConfigDict(extra="forbid")

    device_id: UUID
    asset_id: UUID
    name: str
    role: str


class NetworkPathRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    target_device_id: UUID
    nodes: list[NetworkPathNode]
    connection_ids: list[UUID]
    warnings: list[str]
    documented_path: bool = True


class WorkloadPort(BaseModel):
    model_config = ConfigDict(extra="forbid")

    container_port: int = Field(ge=1, le=65535)
    host_port: int | None = Field(default=None, ge=1, le=65535)
    protocol: str = Field(default="tcp", pattern="^(tcp|udp)$")


class WorkloadUrls(BaseModel):
    model_config = ConfigDict(extra="forbid")

    internal: AnyHttpUrl | None = None
    external: AnyHttpUrl | None = None
    administrative: AnyHttpUrl | None = None
    api: AnyHttpUrl | None = None


class ServiceWorkloadWrite(BaseModel):
    model_config = ConfigDict(extra="forbid")

    host_asset_id: UUID
    name: str = Field(min_length=1, max_length=200)
    image: str | None = Field(default=None, max_length=500)
    image_tag: str | None = Field(default=None, max_length=200)
    compose_project: str | None = Field(default=None, max_length=200)
    network_mode: str = Field(
        default="bridge",
        pattern="^(bridge|host|macvlan|docker_network)$",
    )
    macvlan_address: str | None = Field(default=None, max_length=64)
    ports: list[WorkloadPort] = Field(default_factory=list, max_length=100)
    urls: WorkloadUrls = Field(default_factory=WorkloadUrls)
    reverse_proxy: str | None = Field(default=None, max_length=500)
    dependency_ids: list[UUID] = Field(default_factory=list, max_length=100)
    status: str = Field(default="unknown", pattern="^(running|stopped|planned|unknown)$")
    notes: str | None = Field(default=None, max_length=20_000)

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        return value.strip()

    @model_validator(mode="after")
    def validate_network_mode(self) -> "ServiceWorkloadWrite":
        if self.network_mode != "macvlan" and self.macvlan_address is not None:
            raise ValueError("Eine eigene IP ist nur im Macvlan-Modus zulässig")
        if len(self.dependency_ids) != len(set(self.dependency_ids)):
            raise ValueError("Dienstabhängigkeiten dürfen nicht doppelt vorkommen")
        return self


class ServiceWorkloadRead(ServiceWorkloadWrite):
    id: UUID
    host_name: str
    docker_container_id: str | None = None
    docker_status_text: str | None = None
    docker_networks: list[str] = Field(default_factory=list)
    docker_mounts: list[str] = Field(default_factory=list)
    docker_last_seen_at: datetime | None = None
    docker_managed: bool = False
    created_at: datetime
    updated_at: datetime


class DockerSyncSettingWrite(BaseModel):
    model_config = ConfigDict(extra="forbid")

    enabled: bool = False
    socket_path: str = Field(default="/var/run/docker.sock", min_length=1, max_length=500)
    host_asset_id: UUID | None = None
    refresh_interval_seconds: int = Field(default=300)

    @field_validator("socket_path")
    @classmethod
    def normalize_socket_path(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized.startswith("/"):
            raise ValueError("Der Docker-Socket muss als absoluter Pfad angegeben werden")
        return normalized

    @field_validator("refresh_interval_seconds")
    @classmethod
    def validate_interval(cls, value: int) -> int:
        if value not in {0, 30, 60, 300, 900, 1800}:
            raise ValueError("Ungültiges Docker-Aktualisierungsintervall")
        return value


class DockerSyncSettingRead(DockerSyncSettingWrite):
    host_name: str | None = None
    last_attempt_at: datetime | None = None
    last_success_at: datetime | None = None
    last_error: str | None = None


class DockerSyncResultRead(BaseModel):
    imported: int = Field(ge=0)
    updated: int = Field(ge=0)
    missing: int = Field(ge=0)
    total: int = Field(ge=0)
    docker_version: str | None = None
    synchronized_at: datetime


class DockerConnectionTestRead(BaseModel):
    success: bool
    message: str
    docker_version: str | None = None


class ExportManifestRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    app_version: str
    export_version: str
    created_at: datetime
    modules: list[str]
    excluded_security_fields: list[str]


class ImportPreviewRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    format: str
    export_version: str | None
    record_counts: dict[str, int]
    conflicts: list[str]
    warnings: list[str]
    writable: bool = False


class ImportResultRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    created: int
    skipped: int
    conflicts: int
    modules: list[str]
    rolled_back: bool = False


class AuditEventRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: UUID
    object_type: str
    object_id: str
    object_label: str | None = None
    object_route: str | None = None
    action: str
    change: dict[str, JsonValue]
    display_change: dict[str, JsonValue] | None = None
    created_at: datetime


class GuidedSetupDraftWrite(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=200)
    current_step: int = Field(default=1, ge=1, le=11)
    data: dict[str, JsonValue] = Field(default_factory=dict)

    @field_validator("name")
    @classmethod
    def normalize_draft_name(cls, value: str) -> str:
        return value.strip()


class GuidedSetupDraftRead(GuidedSetupDraftWrite):
    id: UUID
    status: str
    created_at: datetime
    updated_at: datetime


class GuidedSetupPreviewRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    draft_id: UUID
    actions: list[str]
    warnings: list[str]
    errors: list[str]
    duplicate_asset_ids: list[UUID]
    can_apply: bool


class GuidedSetupApplyRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    draft_id: UUID
    asset_id: UUID
    created_object_ids: list[UUID]
    applied_at: datetime


class FritzBoxDeviceRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    mac_address: str | None
    ipv4: str | None
    ipv6: str | None
    active: bool
    interface_type: str | None
    connection_rate_mbps: int | None
    connected_via: str | None
    last_seen: datetime | None
    dhcp_reservation: bool | None

File: frontend/src/types/release.ts
export interface DashboardCardSetting {
  id: 'documentation' | 'consumption_comparison' | 'maintenance' | 'quality' | 'network'
  visible: boolean
}

export interface DashboardSetting {
  cards: DashboardCardSetting[]
  updated_at: string
}

export interface AuditEvent {
  id: string
  object_type: string
  object_id: string
  object_label: string | null
  object_route: string | null
  action: string
  change: Record<string, unknown>
  display_change: Record<string, unknown> | null
  created_at: string
}


export interface FritzBoxDevice {
  name: string
  mac_address: string | null
  ipv4: string | null
  ipv6: string | null
  active: boolean
  interface_type: string | null
  connection_rate_mbps: number | null
  connected_via: string | null
  last_seen: string | null
  dhcp_reservation: boolean | null
}

export interface ImportPreview {
  format: string
  export_version: string | null
  record_counts: Record<string, number>
  conflicts: string[]
  warnings: string[]
  writable: false
}

export interface ImportResult {
  created: number
  skipped: number
  conflicts: number
  modules: string[]
  rolled_back: boolean
}

export interface GuidedSetupDraftWrite {
  name: string
  current_step: number
  data: Record<string, unknown>
}

export interface GuidedSetupDraft extends GuidedSetupDraftWrite {
  id: string
  status: 'draft' | 'applied'
  created_at: string
  updated_at: string
}

export interface GuidedSetupPreview {
  draft_id: string
  actions: string[]
  warnings: string[]
  errors: string[]
  duplicate_asset_ids: string[]
  can_apply: boolean
}

export interface GuidedSetupApply {
  draft_id: string
  asset_id: string
  created_object_ids: string[]
  applied_at: string
}

export interface PortGroupWrite {
  group: 'copper' | 'sfp' | 'sfp_plus' | 'uplink'
  count: number
  scheme: 'numeric' | 'gigabit' | 'ethernet'
  start: number
  speed_mbps: number | null
  poe_capable: boolean
}

export interface PortGenerationPreview {
  device_id: string
  existing_names: string[]
  create_names: string[]
  unchanged_names: string[]
  requested_total: number
  created?: number
}

export interface NetworkPath {
  target_device_id: string
  nodes: Array<{ device_id: string; asset_id: string; name: string; role: string }>
  connection_ids: string[]
  warnings: string[]
  documented_path: true
}

export interface ServiceWorkloadWrite {
  host_asset_id: string
  name: string
  image: string | null
  image_tag: string | null
  compose_project: string | null
  network_mode: 'bridge' | 'host' | 'macvlan' | 'docker_network'
  macvlan_address: string | null
  ports: Array<{ container_port: number; host_port: number | null; protocol: 'tcp' | 'udp' }>
  urls: {
    internal: string | null
    external: string | null
    administrative: string | null
    api: string | null
  }
  reverse_proxy: string | null
  dependency_ids: string[]
  status: 'running' | 'stopped' | 'planned' | 'unknown'
  notes: string | null
}

export interface ServiceWorkload extends ServiceWorkloadWrite {
  id: string
  host_name: string
  docker_container_id: string | null
  docker_status_text: string | null
  docker_networks: string[]
  docker_mounts: string[]
  docker_last_seen_at: string | null
  docker_managed: boolean
  created_at: string
  updated_at: string
}

export interface DockerSyncSettingWrite {
  enabled: boolean
  socket_path: string
  host_asset_id: string | null
  refresh_interval_seconds: 0 | 30 | 60 | 300 | 900 | 1800
}

export interface DockerSyncSetting extends DockerSyncSettingWrite {
  host_name: string | null
  last_attempt_at: string | null
  last_success_at: string | null
  last_error: string | null
}

export interface DockerSyncResult {
  imported: number
  updated: number
  missing: number
  total: number
  docker_version: string | null
  synchronized_at: string
}

export interface DockerConnectionTest {
  success: boolean
  message: string
  docker_version: string | null
}
